Automattic/sensei
# Refactor course video progression logic to work with required blocks

Fixes [sensei-pro#1131](1131-gh-Automattic/sensei-pro)

### Changes proposed in this Pull Request

* Requires all the videos to be played before enabling the Complete Lesson button.
* Autocompletes lesson only when the last "completed" block is video.
* If `window.sensei.store.blocks` redux store is present then delegates the Complete Lesson button logic to it.

### Testing instructions



* For Sensei LMS only
  * Checkout branch `refactor/required-blocks-video-compatibility` for Sensei LMS. Build assets.
  * Add some video blocks into a lesson and save it.
  * Confirm that with **Course Video Progression -> Required** turned on the "Complete Lesson" button is disabled until all the videos in the lesson are played till the end.
  * Confirm that with **Course Video Progression -> Autocomplete Lesson** turned on it auto completes only when the all videos are played till the end.
  * Confirm that with **Course Video Progression -> Autopause** turned on the videos are paused when user navigates out of the current browser tab.
* For Sensei LMS + Sensei Pro
  * Setup Sensei Pro plugin for your installlation. Checkout branch [`feature/required-blocks`](https://github.com/Automattic/sensei-pro/tree/feature/required-blocks). Build assets.
  * Add TaskList block into your the above lesson and mark it as required (see video below hot it's done.)
  * Save the lesson and visit the lesson page.
  * Confirm that the "Complete Lesson" button now has a counter that tells how many videos and blocks needs to be completed.
  * Confirm that the "Complete Lesson" button is enabled only after all the videos are played and all the tasks in tasklist block are completed.


### New/Updated Hooks

* **New Actions**
  * `sensei.videoProgression.registerVideo` Called when a required video for the current lesson is registered.
    * `@param {Object} video`
    * `@param {string} video.url` The source url of the video.
  * `sensei.videoProgression.videoEnded` Called when a required video for the current lesson is finished playing.
    * `@param {Object} video`
    * `@param {string} video.url` The source url of the video.
* **New Filters**
  * `sensei.videoProgression.allCompleted` Tells if all the required videos for the current lesson are finished playing or not.
    * `@param {boolean} allCompleted` Whether all the required videos for the current lesson are completed.
  * `sensei.videoProgression.preventLessonCompletion` Whether or not the Lesson Complete button should be disabled or not.
    * `@param {boolean} shouldPrevent` Whether to prevent users from completing the lesson.
  * `sensei.videoProgression.allowLessonCompletion` Whether or not the Lesson Complete button should be enabled or not.
    * `@param {boolean} shouldAllow` Whether to allow users to complete the lesson.


### Screenshot / Video

#### Course Video Progression Regular Mode

https://user-images.githubusercontent.com/2578542/167613523-48a681da-843a-46d8-8f23-7f6983a101c2.mp4

#### Course Video Progression Learning Mode

https://user-images.githubusercontent.com/2578542/167613589-50da10be-b17d-4d6d-9aa1-cb9efb54f330.mp4

#### Required Blocks Learning Mode

https://user-images.githubusercontent.com/2578542/167613701-cc3507e8-8b25-4046-930e-bfb9867b9c7e.mp4
